core/src/main/java/org/apache/cloudstack/direct/download/HttpsDirectTemplateDownloader.java

```diff
@@ -0,0 +1,252 @@
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class HttpsDirectTemplateDownloader extends DirectTemplateDownloaderImpl {
+
+    protected CloseableHttpClient httpsClient;
+    private HttpUriRequest req;
+
+    protected HttpsDirectTemplateDownloader(String url) {
+        this(url, null, null, null, null, null, null, null, null);
+    }
+
+    public HttpsDirectTemplateDownloader(String url, Long templateId, String destPoolPath, String checksum, Map<String, String> headers,
+                                         Integer connectTimeout, Integer soTimeout, Integer connectionRequestTimeout, String temporaryDownloadPath) {
+        super(url, destPoolPath, templateId, checksum, temporaryDownloadPath);
+        SSLContext sslcontext = getSSLContext();
+        SSLConnectionSocketFactory factory = new SSLConnectionSocketFactory(sslcontext, SSLConnectionSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER);
+        RequestConfig config = RequestConfig.custom()
+                .setConnectTimeout(connectTimeout == null ? 5000 : connectTimeout)
+                .setConnectionRequestTimeout(connectionRequestTimeout == null ? 5000 : connectionRequestTimeout)
+                .setSocketTimeout(soTimeout == null ? 5000 : soTimeout).build();
+        httpsClient = HttpClients.custom().setSSLSocketFactory(factory).setDefaultRequestConfig(config).build();
+        createUriRequest(url, headers);
+        String downloadDir = getDirectDownloadTempPath(templateId);
+        File tempFile = createTemporaryDirectoryAndFile(downloadDir);
+        setDownloadedFilePath(tempFile.getAbsolutePath());
+    }
+
+    protected void createUriRequest(String downloadUrl, Map<String, String> headers) {
+        req = new HttpGet(downloadUrl);
+        if (MapUtils.isNotEmpty(headers)) {
+            for (String headerKey: headers.keySet()) {
+                req.setHeader(headerKey, headers.get(headerKey));
+            }
+        }
+    }
+
+    private SSLContext getSSLContext() {
```

Comment:
there are nested `try` statements in this method. I think it can do with some restructuring.